filebeat/input/kafka/input.go

```diff
@@ -0,0 +1,285 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package kafka
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/Shopify/sarama"
+
+	"github.com/elastic/beats/filebeat/channel"
+	"github.com/elastic/beats/filebeat/input"
+	"github.com/elastic/beats/libbeat/beat"
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/logp"
+
+	"github.com/pkg/errors"
+)
+
+func init() {
+	err := input.Register("kafka", NewInput)
+	if err != nil {
+		panic(err)
+	}
+}
+
+// Input contains the input and its config
+type kafkaInput struct {
+	config        kafkaInputConfig
+	rawConfig     *common.Config // The Config given to NewInput
+	context       input.Context
+	outlet        channel.Outleter
+	consumerGroup sarama.ConsumerGroup
+	sessionState  *kafkaSessionState
+	log           *logp.Logger
+	runOnce       sync.Once
+}
+
+// A synchronized wrapper to read and write the kafka session, since it may
+// change while ACKs are still pending.
+type kafkaSessionState struct {
+	session   sarama.ConsumerGroupSession
+	mutex     sync.Mutex     // Hold to access the session field
+	waitGroup sync.WaitGroup // Hold while using the session field
+}
+
+// NewInput creates a new kafka input
+func NewInput(
+	cfg *common.Config,
+	connector channel.Connector,
+	inputContext input.Context,
+) (input.Input, error) {
+
+	// We create the empty session state first because it must be referenced by
+	// the ACK callback in the connector configuration.
+	sessionState := &kafkaSessionState{}
+
+	out, err := connector.ConnectWith(cfg, beat.ClientConfig{
+		Processing: beat.ProcessingConfig{
+			DynamicFields: inputContext.DynamicFields,
+		},
+		ACKEvents: func(events []interface{}) {
+			sessionState.accessSession(func(session sarama.ConsumerGroupSession) {
+				if session == nil {
+					// The kafka connection is closed and / or is being rebalanced.
+					return
+				}
+				for _, event := range events {
+					if cm, ok := event.(*sarama.ConsumerMessage); ok {
+						session.MarkMessage(cm, "")
+					}
+				}
+			})
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
```

Comment:
The ACKEvents handler can be called for as long as the underlying `beat.Client` is still connected. I guess there is a high chance that the sarama user group closes before the `beat.Client`, giving us ACKs for a session that has been closed already. Looks like this is the also a reason you have to check for `nil` session for every event.

I wonder if the session is bound to a handler. In this case 'Private' could be a tuple of handler+event. All logic about tracking the liveness of a session + ACKing messages would be local to the handler then, reducing the need for some 'global' sharing across instances of different types.

e.g.

```
type eventExtra struct {
  handler *groupHandler
  message *sarame.ConsumerMessage
}

...

ACKHandler: func(privates []interface{}) {
  for _, priv := range privates {
    if extra, ok := priv.(eventExtra); ok {
      extra.handler.ACK(extra.message)
    }
  }
}
```

WDYT?

The ClientConfig itself supports a 'WaitShutdown' setting, to postpone shutdown for a little. If configured `Close` will wait until all events have been ACKed or after until the configured timeout. It would be nice to have the same functionality via a custom delaying context and expose this functionality via configs to users. This would reduce the chance of duplicates incurred by shutdown. We can add this functionality in a later PR.

Reply:
The groupHandler seems to be 'our' side of the session. If we protect Start, Cleanup, and ACK methods using a mutex, then we would have the session sync, handling + acking in one type only.

Reply:
One thing I wasn't confident of on the first pass, but I realize you've since confirmed, is that ACKs will be sent sequentially, so that in particular only one ACK can be executed at a time (so the wait group is superfluous and the whole system can operate on a single mutex). Combined with the mistake mentioned in another comment about network errors returned from `NewConsumerGroup`, the run loop / ack logic needs revision. I will get you a simpler and more correct version soon.